function-server/src/main/java/com/github/dispatchframework/javabaseimage/Server.java

```diff
@@ -0,0 +1,149 @@
+        Undertow.builder()
+                .addHttpListener(8080, "0.0.0.0")
+                .setHandler(Handlers.path()
+                        .addPrefixPath("/", new ExecFunction(args[0], args[1]))
+                        .addExactPath("/healthz", new Healthz()))
+                .build()
+                .start();
+    }
+
+    private static class ExecFunction implements HttpHandler {
+        private String packageName;
+        private String className;
+        private final Gson gson;
+
+        public ExecFunction(String packageName, String className) {
+            this.packageName = packageName;
+            this.className = className;
+            this.gson = new GsonBuilder().serializeNulls().create();
+        }
+
+        @Override
+        public void handleRequest(final HttpServerExchange exchange) {
+            exchange.getRequestReceiver().receiveFullString((httpServerExchange, message) -> {
+                Object r = null;
+                Exception err = null;
+
+                // Closing a ByteArrayOutputStream has no effect
+                ByteArrayOutputStream baosStderr = new ByteArrayOutputStream();
+                ByteArrayOutputStream baosStdout = new ByteArrayOutputStream();
+
+                PrintStream oldStderr = System.err;
+                PrintStream oldStdout = System.out;
+
+                try (PrintStream stderr = new PrintStream(baosStderr);
+                     PrintStream stdout = new PrintStream(baosStdout)) {
+                    try {
+                        System.setErr(stderr);
+                        System.setOut(stdout);
+
+                        Class<?> functionClass = Class.forName(packageName + "." + className);
+                        Type[] genericInterfaces = functionClass.getGenericInterfaces();
+                        Type[] genericTypes = null;
+
+                        for (Type genericInterface : genericInterfaces) {
+                            if (genericInterface instanceof ParameterizedType) {
+                                ParameterizedType parameterizedType = (ParameterizedType) genericInterface;
+                                if (parameterizedType.getRawType().equals(BiFunction.class)) {
+                                    genericTypes = parameterizedType.getActualTypeArguments();
+                                    break;
+                                }
+                            }
+                        }
+
+                        JsonObject rootObj = new JsonParser().parse(message).getAsJsonObject();
+                        String context = rootObj.get("context").toString();
+                        String payload = rootObj.get("payload").toString();
+
+                        BiFunction f = (BiFunction) functionClass.newInstance();
+
+                        if (genericTypes != null && genericTypes.length >= 2) {
+                            r = f.apply(gson.fromJson(context, genericTypes[0]), gson.fromJson(payload, genericTypes[1]));
```

Comment:
Looks like context and payload are going through this process before being passed to the function:
1. parse the request from JSON
2. print context and payload to JSON separately
3. parse both context and payload from JSON again

Can this be done more efficiently? 

Also, looks like this is a great candidate for factoring out as a separate function and unit testing

Reply:
I don't think there is much of a way to make it more efficient. Right now the request is not initially being parsed completely. It is only grabbing the root of the request which it then separates to the context and payload JSON representations without decoding the underlying structure. Then when calling f.apply, it decodes the context and payload to its actual object representations.

Without being given the concrete types of the context and payload we can't directly parse the request into an object with a context and payload. We first have to separate out the request into the context and payload JSON parts and then parse them separately into java objects.

Reply:
I don't think there is much of a way to make it more efficient. Right now the request is not initially being parsed completely. It is only grabbing the root of the request which it then separates to the context and payload JSON representations without decoding the underlying structure. Then when calling f.apply, it decodes the context and payload to its actual object representations.

Without being given the concrete types of the context and payload we can't directly parse the request into an object with a context and payload. We first have to separate out the request into the context and payload JSON parts and then parse them separately into java objects.